dlt/common/plugins/plugins.py

```diff
@@ -0,0 +1,143 @@
+from typing import List, Any, Iterable, Optional, Dict
+from dlt.common.typing import TFun
+from dlt.common.typing import TDataItem
+from dlt.common.pipeline import SupportsPipeline
+from dlt.common.configuration.specs import ContainerInjectableContext
+from dlt.common.configuration.specs import configspec
+from functools import wraps
+from .reference import SupportsCallbackPlugin, Plugin, TSinglePluginArg, TPluginArg, CallbackPlugin
+from dlt.common.schema.exceptions import DataValidationError
+from importlib import import_module
+from .exceptions import UnknownPluginPathException
+from dlt.common.configuration.specs.base_configuration import BaseConfiguration
+import multiprocessing as mp
+from functools import wraps
+import threading
+import queue
+
+
+SAFE_SUFFIX = "_safe"
+
+
+def create_safe_version(f: TFun) -> TFun:
+    @wraps(f)
+    def _wrap(self: "PluginsContext", *args: Any, **kwargs: Any) -> Any:
+        # send message to shared queue if this is not the main instance
+        if self._plugins:
+            if not self._main or threading.main_thread() != threading.current_thread():
+                self._queue.put((f.__name__, args, kwargs))
+            else:
+                getattr(self, (f.__name__ + SAFE_SUFFIX))(*args, **kwargs)
+        return f(self, *args, **kwargs)
+
+    return _wrap  # type: ignore
+
+
+@configspec
+class PluginsContext(ContainerInjectableContext, SupportsCallbackPlugin):
+    def __init__(self, main: bool = True) -> None:
+        self._plugins: List[Plugin[BaseConfiguration]] = []
+        self._callback_plugins: List[CallbackPlugin[BaseConfiguration]] = []
+        self._initial_plugins: TPluginArg = []
+        self._main = main
+        self._queue: Optional[queue.Queue[Any]] = None
```

Comment:
I think here you might potentially have linting issues because `queue.Queue[Any]` is not providing generic way of doing thus mypy might complaint that `Queue is not subscriptable....`
So I did the following workaround
```py
# Please read more at https://mypy.readthedocs.io/en/stable/runtime_troubles.html#not-generic-runtime
T = TypeVar("T")
if TYPE_CHECKING:

    class _Queue(Queue[T]):
        pass

else:

    class _Queue(Generic[T], Queue):
        pass


class BaseQueue(_Queue[T]):
    pass
```